### PR TITLE
test : getSecurityGroupName()

### PR DESCRIPTION
**Pull request by KingDaemonX**
**What this PR does / why we need it**:

**Which issue this PR fixes(if applicable)**:
fixes #

**Special notes for reviewers**:

**Release note**:

```release-note
NONE
```
implement a test case that checks for edge case in the `getSecurityGroupName` function to testing all logic and hardcode to check for changes in the function code

### CODE REVIEW

**Comment by dulek** [comment]
/ok-to-test

Oh, there's a conflict here as another patch merged. You'll need to solve it.

File: pkg/openstack/loadbalancer_test.go [hunk 1]
```diff
@@ -460,3 +462,50 @@ func TestGetRulesToCreateAndDelete(t *testing.T) {
...
+		{
+			name: "security group name longer than 255 byte",
+			service: &corev1.Service{
+				ObjectMeta: v1.ObjectMeta{
+					UID:       "12345678-90ab-cdef-0123-456789abcdef",
+					Namespace: "security-group-longer-test-namespace",
+					Name:      "security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end-e-end-pardon-the-long-string-i-really-apologize-if-this-is-a-bad-thing-to-do",
+				},
+			},
+			expect: "lb-sg-12345678-90ab-cdef-0123-456789abcdef-security-group-longer-test-namespace-security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end",
```
[the remark this answers is not in the document]
Comment (KingDaemonX) [reply pkg/openstack/loadbalancer_test.go:492]: yeah, it but i had it hardcoded, incase of unexpected changes

File: pkg/openstack/loadbalancer_test.go [hunk 2]
```diff
@@ -534,3 +537,51 @@ func Test_getListenerProtocol(t *testing.T) {
...
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			test.expect = fmt.Sprintf("lb-sg-%s-%s-%s", test.service.UID, test.service.Namespace, test.service.Name)
+			if len(test.expect) > 255 {
+				test.expect = test.expect[:255]
+			}
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:580]: We shouldn't compute the values in the test runtime. They should be hardcoded, so that we notice any change that affects what `getSecurityGroupName()` generates.

E.g. for the third testcase, the value of `expect` should be `expect: lb-sg---`.

**Review by dulek — CHANGES_REQUESTED** [review]

**Review by dulek — CHANGES_REQUESTED** [review]

File: pkg/openstack/loadbalancer_test.go [hunk 3]
```diff
@@ -534,3 +536,30 @@ func Test_getListenerProtocol(t *testing.T) {
...
+	service := &corev1.Service{ObjectMeta: v1.ObjectMeta{UID: "12345", Namespace: "security-group-namespace", Name: "security-group-name"}}
+	expected := "lb-sg-12345-security-group-namespace-security-group-name"
+	t.Run("regular test security group name and length", func(t *testing.T) {
+		if got := getSecurityGroupName(service); got != expected {
+			t.Errorf("getSecurityGroupName() = %s expected = %s", got, expected)
+		}
+	})
+
+	service = &corev1.Service{ObjectMeta: v1.ObjectMeta{UID: "12345678-90ab-cdef-0123-456789abcdef", Namespace: "security-group-longer-test-namespace", Name: "security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end-e-end-pardon-the-long-string-i-really-apologize-if-this-is-a-bad-thing-to-do"}}
+	expected = "lb-sg-12345678-90ab-cdef-0123-456789abcdef-security-group-longer-test-namespace-security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end"
```
Comment (sakshi-1505) [inline pkg/openstack/loadbalancer_test.go:551]: Let's make sure the tests are readable with proper lining & spacing in the object declarations.

File: pkg/openstack/loadbalancer_test.go [hunk 4]
```diff
@@ -534,3 +536,30 @@ func Test_getListenerProtocol(t *testing.T) {
...
+	t.Run("security group name longer than 255 byte", func(t *testing.T) {
+		if got := getSecurityGroupName(service); got != expected {
+			t.Errorf("getSecurityGroupName() = %s expected = %s", got, expected)
+		}
+	})
+
+	service = &corev1.Service{}
+	expected = "lb-sg---"
```
Comment (sakshi-1505) [inline pkg/openstack/loadbalancer_test.go:559]: Any reason they are being initialised outside of the test block within `t.Run`?

File: pkg/openstack/loadbalancer_test.go [hunk 5]
```diff
@@ -534,3 +536,30 @@ func Test_getListenerProtocol(t *testing.T) {
...
+
+func Test_getSecurityGroupName(t *testing.T) {
+
+	service := &corev1.Service{ObjectMeta: v1.ObjectMeta{UID: "12345", Namespace: "security-group-namespace", Name: "security-group-name"}}
+	expected := "lb-sg-12345-security-group-namespace-security-group-name"
+	t.Run("regular test security group name and length", func(t *testing.T) {
+		if got := getSecurityGroupName(service); got != expected {
+			t.Errorf("getSecurityGroupName() = %s expected = %s", got, expected)
+		}
+	})
```
Comment (sakshi-1505) [inline pkg/openstack/loadbalancer_test.go:548]: as @dulek suggested, let's try to stick to DRY(don't repeat yourself) principle rather than initialising the run block with mock values every time.

**Review by sakshi-1505 — CHANGES_REQUESTED** [review]

File: pkg/openstack/loadbalancer_test.go [hunk 6]
```diff
@@ -534,3 +536,30 @@ func Test_getListenerProtocol(t *testing.T) {
...
+
+	service = &corev1.Service{ObjectMeta: v1.ObjectMeta{UID: "12345678-90ab-cdef-0123-456789abcdef", Namespace: "security-group-longer-test-namespace", Name: "security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end-e-end-pardon-the-long-string-i-really-apologize-if-this-is-a-bad-thing-to-do"}}
+	expected = "lb-sg-12345678-90ab-cdef-0123-456789abcdef-security-group-longer-test-namespace-security-group-longer-test-service-name-with-more-than-255-byte-this-test-should-be-longer-than-255-i-need-that-ijiojohoo-afhwefkbfk-jwebfwbifwbewifobiu-efbiobfoiqwebi-the-end"
+	t.Run("security group name longer than 255 byte", func(t *testing.T) {
+		if got := getSecurityGroupName(service); got != expected {
+			t.Errorf("getSecurityGroupName() = %s expected = %s", got, expected)
+		}
+	})
```
Comment (sakshi-1505) [inline pkg/openstack/loadbalancer_test.go:556]: Let's use assert package here too
